### DISCUSSION TITLE
🧪🔬 Early Experiments & Learnings: Testing Lunaris Codex's Core Learning Capability

### DISCUSSION THREAD

**Discussion by MeryylleA** [MeryylleA/lunariscodex#38] [Show and tell]
Hey everyone, and welcome to the Lunaris Codex project!

As the primary developer ([@MeryylleA](https://github.com/MeryylleA)), I'm incredibly excited about building this open-source PyTorch Transformer Decoder framework. Our goal is to provide a flexible, understandable, and powerful toolkit for anyone interested in language modeling and code generation. You can find the main repository here: [https://github.com/MeryylleA/lunariscodex](https://github.com/MeryylleA/lunariscodex)

We've recently been conducting some foundational experiments to validate the core learning capabilities of the Lunaris Codex architecture and its training pipeline, especially within a resource-constrained environment (primarily CPU, with some initial tests on a free-tier Colab GPU). I wanted to share some of these early, raw, and often amusing results, as they highlight both the potential of the framework and the realities of training LLMs.

**Why These "Toy" Experiments?**

Before attempting large-scale training runs (which require significant data and compute), it's crucial to ensure the fundamental pieces are working:
1.  **Can the model learn at all?** Does loss decrease? Does it overfit to simple patterns?
2.  **Is the data pipeline robust?** Can it process and tokenize data correctly?
3.  **Is the training loop functioning as expected?** Are gradients flowing, optimizers updating, and checkpoints saving?
4.  **How does the model behave with different tokenizers and data types?**

These initial tests are "sanity checks" for the framework itself.

**Experiment 1: The "2+2=4" Overfitting Test**

*   **Dataset:** 10,000 lines of "2+2=4", plus a small validation set.
*   **Model:** Tiny Lunaris Codex (d_model=64, 2 layers, 2 heads, ~1.6M params).
*   **Tokenizer:** `gpt2`.
*   **Goal:** See if the model could perfectly memorize this single, repeated pattern.
*   **Result:** **Success!** The model quickly overfit, achieving near-zero loss, PPL close to 1.0, and 100% accuracy. Inference with `prompt="2+2="` correctly (and amusingly, if allowed more tokens) generated `4` and then continued with more `4`s.
    *   *Log Snippet (CPU Training):*
        ```
        Epoch 10/10 [Training]: ... loss=0.0001, ppl=1.00, acc=1.000
        [Validation/Test]: ... val_loss=0.0001, val_ppl=1.00, val_acc=1.000
        ```
    *   *Inference (Prompt "2+2=", max_new_tokens 100):* `4444...`
*   **Key Takeaway:** The core architecture and training loop can learn and memorize effectively.

**Experiment 2: Fine-tuning "2+2=4" Model with Other Simple Sums**

*   **Dataset:** ~800 lines of various simple sums (e.g., "1+1=2", "3+5=8"), with repetitions.
*   **Model:** The checkpoint from Experiment 1.
*   **Goal:** See if the model could learn new sums and if it exhibited catastrophic forgetting.
*   **Result:** Mixed and very interesting!
    *   The model *partially* retained "2+2=4" (e.g., `prompt="2+2="` -> `4=6`).
    *   It struggled to correctly learn new sums, often mixing numbers or defaulting to common results like "=10" (e.g., `prompt="3+5="` -> `7=10`).
    *   This highlighted the challenge of learning even simple arithmetic/pattern matching with a tiny model and limited, skewed data.
*   **Key Takeaway:** Shows how sensitive small models are to data distribution and the difficulty of true generalization without more capacity and diverse data.

**Experiment 3: Training on "Simple Phrases" with a BERT Tokenizer (CPU & GPU)**

*   **Dataset:** ~5000 lines of common Portuguese phrases (greetings, simple questions), heavily repeated.
*   **Model:** Small Lunaris Codex (d_model=128, 4 layers, 4 heads, ~4.6M params).
*   **Tokenizer:** `neuralmind/bert-base-portuguese-cased`.
*   **Goal:** Test with a different data type (natural language) and tokenizer.
*   **Result (CPU & GPU on Colab Tesla T4):**
    *   Training metrics (loss, PPL, acc) improved significantly on the training/validation set, showing overfitting.
    *   **GPU training was massively faster (e.g., ~1-2 seconds/epoch on Colab GPU vs. ~1 min/epoch on local CPU for a similar setup with a smaller dataset). This validated CUDA and AMP (`fp16`) support!**
    *   Inference results were quite "creative" and often off-topic (e.g., `prompt="bom dia,"` -> `o banheiro???????`).
*   **Key Takeaway:** The framework is GPU-capable! However, a small model on a small, repetitive natural language dataset still defaults to superficial pattern matching rather than deep understanding. The choice of tokenizer also clearly influences behavior.

**Experiment 4: Training on `meryyllebr543/lunaris-data` (Code/Technical Focus - 1 Epoch on CPU)**

*   **Dataset:** 5,000 examples from the `lunaris-data` (English, Python-focused, `USER:/ASSISTANT:` format).
*   **Model:** Slightly larger Lunaris Codex (d_model=256, 6 layers, 4 heads, ~17.7M params).
*   **Tokenizer:** `bigcode/starcoder`.
*   **Goal:** Test with a more realistic (but still small) dataset and a slightly larger model.
*   **Result (after ~3 hours for 1 epoch on CPU!):**
    *   Training loss dropped PLHIV from >10 to an average of ~4.99 (ending epoch batch loss ~3.6).
    *   Inference showed the model attempting to mimic the *structure* of technical documentation/code explanations (e.g., using Markdown headers, ` ```typescript ` blocks, technical jargon).
    *   It did **not** generate functionally correct code for specific prompts yet, often getting stuck on placeholder tokens (like "`n`") or mixing concepts. However, the *attempt* to generate structured, domain-relevant (even if incorrect) output was a clear step up from random noise.
    *   *Example (Prompt: "USER: Write a Python function that adds two numbers.\nASSISTANT:")*
        ```
         ### Cases
        
         ### Step 1: Implementing the `undefined` for **Edge Cases**: The `n` method is a `n` function...
        ```
*   **Key Takeaway:** The architecture *can* process and learn from more complex, code-related data. Even one epoch starts to instill structural patterns. The need for more data, more epochs, and more powerful hardware (or very long training times) for meaningful code generation is evident. The sensitivity to prompt formatting (e.g., including `ASSISTANT:`) was also very clear.

**Overall Learnings & Next Steps for Lunaris Codex:**

*   **The Core Framework Works:** Lunaris Codex can load data, initialize models, train (on CPU & GPU with AMP), save checkpoints, and perform inference. The model architecture is capable of learning and overfitting, which is a crucial baseline.
*   **Data is King (and Queen, and the Entire Royal Court):** These tests starkly demonstrate how heavily model behavior is influenced by the training data's size, diversity, and quality.
*   **Small Models, Big Dreams, Realistic Expectations:** Our "toy" models aren't going to write complex software (yet!), but they are invaluable for validating the framework and understanding LLM fundamentals.
*   **Roadmap Focus:**
    1.  **Integrate our custom `bpe_trainer` into `prepare_data.py`:** Allowing fully custom tokenization is a high priority.
    2.  **Enhance `lunaris_data_analyzer`:** Add token decoding for better data insights.
    3.  **Expand Test Coverage:** More unit tests for `model.py`.
    4.  **Develop `evaluate.py`:** For more standardized model evaluation.

I'm incredibly proud of how far Lunaris Codex has come, and these experiments, while sometimes yielding funny results, are super encouraging. It's clear the foundation is solid.

I'd love to hear your thoughts, suggestions, or if you've had similar experiences with your own LLM experiments! What kind of "first steps" do you usually take when building or testing a new framework?

Cheers,
Francisco ([@MeryylleA](https://github.com/MeryylleA))
(P.S. Big thanks to Google's Gemini for the ongoing pair-programming, brainstorming, and debugging support on this project!)
